private/buf/buflsp/completion.go

```diff
@@ -0,0 +1,533 @@
+	if decl.IsZero() {
+		// No declaration found, return top-level keywords.
+		return topLevelCompletionItems()
+	}
+
+	// Return context-specific completions based on declaration type.
+	switch decl.Kind() {
+	case ast.DeclKindDef:
+		return completionItemsForDef(file, decl.AsDef(), prefix)
+	case ast.DeclKindSyntax:
+		// Inside syntax declaration - could suggest syntax values.
+		return nil
+	case ast.DeclKindPackage:
+		// Inside package declaration - could suggest package name based on path.
+		return nil
+	case ast.DeclKindImport:
+		// Inside import declaration - could suggest importable files.
+		return completionItemsForImport(file)
+	default:
+		// For other declaration types, fall back to top-level keywords.
+		return topLevelCompletionItems()
+	}
+}
+
+// completionItemsForDef returns completion items for definition declarations (message, enum, service, etc.).
+func completionItemsForDef(file *file, def ast.DeclDef, prefix string) []protocol.CompletionItem {
+	switch def.Classify() {
+	case ast.DefKindMessage:
+		// Inside a message - suggest field keywords, types, and nested declarations.
+		return completionItemsForMessage(file, prefix)
+	case ast.DefKindService:
+		// Inside a service - suggest rpc and option keywords.
+		return completionItemsForService()
+	case ast.DefKindEnum:
+		// Inside an enum - suggest option keyword.
+		return []protocol.CompletionItem{
+			keywordToCompletionItem(keyword.Option),
+		}
+	default:
+		// For other definitions, return top-level keywords.
+		return topLevelCompletionItems()
+	}
+}
+
+// completionItemsForMessage returns completion items for inside a message definition.
+func completionItemsForMessage(file *file, prefix string) []protocol.CompletionItem {
+	// Keywords for message body.
+	messageKeywords := []keyword.Keyword{
+		keyword.Message,
+		keyword.Enum,
+		keyword.Option,
+		keyword.Oneof,
+		keyword.Extensions,
+		keyword.Reserved,
+		keyword.Extend,
+		// Field keywords
+		keyword.Repeated,
+		keyword.Optional,
+		keyword.Required,
+	}
+
+	items := xslices.Map(messageKeywords, keywordToCompletionItem)
+
+	// Add predeclared types (primitives).
+	items = append(items, predeclaredTypeCompletionItems()...)
+
+	// Add referenceable types (messages, enums) from this file and imports.
+	items = append(items, referenceableTypeCompletionItems(file, prefix)...)
+
+	return items
+}
+
+// completionItemsForService returns completion items for inside a service definition.
+func completionItemsForService() []protocol.CompletionItem {
+	serviceKeywords := []keyword.Keyword{
+		keyword.RPC,
+		keyword.Option,
+	}
+	return xslices.Map(serviceKeywords, keywordToCompletionItem)
+}
+
+// completionItemsForImport returns completion items for import declarations.
+func completionItemsForImport(file *file) []protocol.CompletionItem {
+	// Suggest importable file paths.
+	paths := xslices.MapKeysToSlice(file.importToFile)
```

Comment:
This allocates an extra slice. I would just iterate over the map instead of using this.